Task BJS2-73472

FROLOV-KI opened the pull request:
1) House class added
2) Room and Food classes added
3) Food collecting method added

FROLOV-KI committed:
Task BJS2-73472

1) House class added
2) Room and Food classes added
3) Food collecting method added

20cognosce commented on src/main/java/school/faang/mices/Room.java:
```diff
@@ -0,0 +1,31 @@
+        this.roomFood = roomFood;
+    }
+
+    public Food peekFood() {
+        synchronized (roomFood) {
+            return roomFood.remove(0);
+        }
+    }
+
+    public boolean isFoodInTheRoom() {
+        synchronized (roomFood) {
+            return !roomFood.isEmpty();
+        }
+    }
```
можно синхронизированные коллекции использовать

20cognosce commented on src/main/java/school/faang/mices/House.java:
```diff
@@ -0,0 +1,63 @@
+    private List<Room> rooms;
+
+    public House(List<Room> rooms) {
+        this.initialRoomsList = rooms;
+        this.rooms = initialRoomsList;
```
в чем отличие? initialRoomsList от rooms?

20cognosce replied:
оставь только один список с текущими, еще не занятыми комнатами
можешь вместо списка обратить внимание на очередь

и при создании объекта просто в очередь положить элементы в перемешанном рандомном порядке
тогда в рантайме тебе не придется ничего с рандомом мутить

FROLOV-KI replied:
Ну это скорее для проверки, первый спискок исходный, я его в конце вывожу, чтобы убедится, что все пусто, так он не нужен

20cognosce commented on src/main/java/school/faang/mices/House.java:
```diff
@@ -0,0 +1,63 @@
+        SecureRandom randomized = new SecureRandom();
+        int firstRoomIndex = randomized.nextInt(0, rooms.size());
+        int secondRoomIndex = randomized.nextInt(0, rooms.size());
+        return firstRoomIndex == secondRoomIndex ? List.of(firstRoomIndex) : List.of(firstRoomIndex, secondRoomIndex);
```
а если рандом вернул одну и ту же комнату, но при этом еще есть не пустые? косяк
я бы заранее распределил комнаты перед стартом

FROLOV-KI replied:
У меня за раз из комнаты убирается 1 элемент еды, а не все, поэтому если двум потокам выдаст одну комнату, то каждый из них заберет по одной еде, а если комната пуста ее из списка убираю

20cognosce commented on src/main/java/school/faang/mices/House.java:
```diff
@@ -0,0 +1,63 @@
+                    .toList();
+        }
+
+        List<Food> pickedFood = new ArrayList<>();
+        peekedRoom.stream()
```
где гарантия, что в этом месте два потока не получили одну и ту же комнату? рандом взял и выдал им одинаковые
не вижу этого

20cognosce commented on src/main/java/school/faang/mices/House.java:
```diff
@@ -0,0 +1,63 @@
+            collectedFood.addAll(pickedFood);
+        }
+
+        synchronized (initialRoomsList) {
+            this.rooms = rooms.stream().filter(Room::isFoodInTheRoom).toList();
+            log.info("{}", rooms);
+        }
```
а зачем это делать? Это слишком не оптимально, нельзя при выдаче сразу убрать команду из списка?

20cognosce commented on src/main/java/school/faang/mices/House.java:
```diff
@@ -0,0 +1,63 @@
+        }
+    }
+
+    private List<Integer> randomRoomIndexes() {
```
почему бы не сделать этот метод синхронизированным и чтобы он сразу выдавал две комнаты, удалив их из списка? зачем индексы какие-то

20cognosce commented on src/main/java/school/faang/mices/House.java:
```diff
@@ -0,0 +1,63 @@
+    public void collectFood() {
+        List<Room> peekedRoom;
+
+        synchronized (initialRoomsList) {
```
ты везде используешь один и тот же способ синхронизации, делай методы синхронизированными, не надо захватывать мониторы чего попало, это было только примером в одной из первых задач

20cognosce commented on src/main/java/school/faang/mices/House.java:
```diff
@@ -0,0 +1,63 @@
+                .filter(Room::isFoodInTheRoom)
+                .forEach(room -> pickedFood.add(room.peekFood()));
+
+        synchronized (collectedFood) {
+            collectedFood.addAll(pickedFood);
+        }
```
попробуй синхронизированную коллекцию

20cognosce reviewed (changes requested):
House и Room доработать

FROLOV-KI committed:
PR issues resolved

1) Concurrent collections added
2) Room peek method refactored
3) Thread creation refactored

FROLOV-KI committed:
PR issues resolved

1) Concurrent collections added
2) Room peek method refactored
3) Thread creation refactored

FROLOV-KI committed:
Merge remote-tracking branch 'origin/kiris_the_best_BJS2-73472' into kiris_the_best_BJS2-73472

# Conflicts:
#	src/main/java/school/faang/mices/House.java
